unit/util/lazy.cpp

```diff
@@ -0,0 +1,29 @@
+/*******************************************************************\
+
+Module: Unit tests for lazy
+
+Author: Romain Brenguier, romain.brenguier@diffblue.com
+
+\*******************************************************************/
+
+#include <testing-utils/use_catch.h>
+#include <util/lazy.h>
+
+SCENARIO("lazy test", "[core][util][lazy]")
+{
+  std::size_t call_counter = 0;
+  auto length_with_counter = [&call_counter](const std::string &s) {
+    ++call_counter;
+    return s.length();
+  };
+  auto lazy_length =
+    lazyt<int>::from_fun([&]() { return length_with_counter("foo"); });
```

Comment:
😊 https://docs.microsoft.com/en-us/dotnet/fsharp/language-reference/lazy-expressions
And now you could even infer return type by doing:
```
template<T> auto lazy(T func) -> decltype(func()) { ... }
```

Reply:
I'm adding it in a follow-up PR: https://github.com/diffblue/cbmc/pull/5101